src/server-glue/webresource-handler.ts

```diff
@@ -0,0 +1,327 @@
+	}
+
+	const allModifiedFields: string[] = Array.isArray(modifiedFields)
+		? modifiedFields.flatMap((mf) => mf?.fields ?? [])
+		: modifiedFields.fields ?? [];
+
+	return allModifiedFields;
+};
+
+const deleteFiles = async (
+	keysToDelete: string[],
+	webResourceHandler: WebResourceHandler,
+) => {
+	const promises = keysToDelete.map((r) => webResourceHandler.removeFile(r));
+	await Promise.all(promises);
+};
+
+const getCreateWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		'POSTRUN-ERROR': async ({ tx, request }) => {
+			tx?.on('rollback', async () => {
+				const fields = getWebResourceFields(request);
+
+				if (fields.length === 0) {
+					return;
+				}
+
+				const keysToDelete: string[] = fields.map(
+					(f) => request.values[f].href,
+				);
+				await deleteFiles(keysToDelete, webResourceHandler);
+			});
+		},
+	} as sbvrUtils.Hooks;
+};
+
+const isDefined = <T>(x: T | undefined | null): x is T => x != null;
+
+const getWebResourcesHrefs = (webResources?: WebResourcesDbResponse[]) => {
+	const resources = webResources ?? [];
+	const hrefs = resources
+		.flatMap((resource) => Object.values(resource ?? {}))
+		.filter(isDefined)
+		.map((resourceKey) => resourceKey.href);
+	return hrefs;
+};
+
+const getRemoveWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		PRERUN: async (args) => {
+			const { api, request } = args;
+			let fields = getWebResourceFields(request);
+
+			if (fields.length === 0) {
+				return;
+			}
+
+			if (request.method === 'PATCH') {
+				const allFields = getModifiedFields(request).map(sqlNameToODataName);
+				fields = fields.filter((f) => allFields.includes(f));
+			}
+
+			if (fields.length === 0) {
+				return;
+			}
+
+			if (request.method === 'PATCH' && request.odataQuery?.key == null) {
+				throw new errors.BadRequestError(
+					'WebResources can only be updated when providing a resource key.',
+				);
+			}
+
+			const ids = await sbvrUtils.getAffectedIds(args);
+			if (ids.length === 0) {
+				return;
+			}
+
+			const webResources = (await api.get({
+				resource: request.resourceName,
+				passthrough: {
+					tx: args.tx,
+					req: permissions.root,
+				},
+				options: {
+					$select: fields,
+					$filter: {
+						id: {
+							$in: ids,
+						},
+					},
+				},
+			})) as WebResourcesDbResponse[] | undefined;
+
+			request.custom.$pineWebResourcesToDelete =
+				getWebResourcesHrefs(webResources);
+		},
+		POSTRUN: ({ tx, request }) => {
+			tx.on('end', () => {
+				const keysToDelete: string[] =
+					request.custom.$pineWebResourcesToDelete || [];
+				// on purpose does not await for this promise to resolve
+				deleteFiles(keysToDelete, webResourceHandler);
+			});
+		},
+	} as sbvrUtils.Hooks;
+};
+
+export const getDefaultHandler = (): WebResourceHandler => {
+	let handler: WebResourceHandler;
+	try {
+		handler = new S3Handler();
+	} catch (e) {
+		console.warn(`Failed to initialize S3 handler, using noop ${e}`);
```

Comment:
@fisehara My understanding here is that S3 dependency/WebResources is an optional feature of Pinejs. And a backwards compactible one. Let's say you run a pine application which will not use WebResources. You might not want to setup a minio/s3 bucket at all. If ever comes a time where you need it, you add the S3 env variables and it will use S3 handler.

Reply:
@otaviojacobi 
Yeah that's okay but if you have defined it, it should load and not fall back to a noop handler. If you don't need it it should not initialise anything, right?

Reply:
@fisehara  The problem is: Previously, we used multer to handle multi-part requests (either they constained a file or not). Now, we are handling multi-part requests ourselves with Busboy library. To keep the same behaviour as before, even if a file is passed and ignored, busboy requires you to at least close the handle (if you look into the Noop, it just does that: ignores the file), maybe renaming into IgnoreFile handler? Because it is just there to keep multi-part requests support (even if they don't have files)